projects/library/src/lib/states/array.state.ts

```diff
@@ -7,217 +7,235 @@ import { BaseState } from './base.state';
-
-    protected override setNewValue(value: T[] | null): void {
-        if (value) {
-            super.setNewValue([...value]);
-        } else {
-            super.setNewValue(null);
-        }
-    }
-
-    protected override catchError(error: Error | TypeError, actionName: string): void {
-        if (error instanceof TypeError) {
-            throw new Error(
-                `\n${this.constructor.name} [${actionName}]: ` +
-                `Firstly set Array.\n\n${error.message}`
-            );
-        }
-
-        super.catchError(error, actionName);
-    }
-
-    /**
-     *	Must return identify param of item.
-     *	Method must be filled in child classes.
-     *	Used for compare two any items.
-     *  @protected
-     *	@param {Generic} item - item of your state.
-     *	@return {Generic} identify param of item.
-     */
-    protected getItemId(item: T): any {
-        return item;
-    }
-
-    /**
-     * 	Compare two items via `getItemId`
-     *	@private
-     *	@param {Generic} itemToUpdate - item that will be update.
-     *	@return {boolean} result of comparing two items via `getItemId`.
-     */
-    private compareItems(firstItem: T, secondItem: T): boolean {
-        return (this.getItemId(firstItem) === this.getItemId(secondItem));
-    }
+  /**
+   * 	Return item by quired index.
+   *	@public
+   *	@param {Number} index - Quired index
+   *  @deprecated use `this.data[index]` instead
+   *	@return {Generic} quired item.
+   */
+  public getByIndex(index: number): T | undefined {
+    const items = this.data;
+
+    return items![index];
+  }
+
+  /**
+   * 	Unshift item to array in state.
+   *	@public
+   *	@param {Generic} item - Item needs to unshift.
+   */
+  @Action
+  public unshiftItem(item: T): void {
+    const items = this.data;
+
+    items!.unshift(item);
+
+    this.setNewValue(items);
+  }
+
+  /**
+   * 	Shift array in state.
+   *	@public
+   */
+  @Action
+  public shift(): void {
+    const items = this.data;
+
+    items!.shift();
+
+    this.setNewValue(items);
+  }
+
+  /**
+   * 	Pop array in state.
+   *	@public
+   */
+  @Action
+  public pop(): void {
+    const items = this.data;
+
+    items!.pop();
+
+    this.setNewValue(items);
+  }
+
+  /**
+   * 	Concat current state with another array.
+   *	@param {T[]} array - Another array to concat with the current state.
+   *	@public
+   */
+  @Action
+  public concatWith(array: T[]): void {
+    const items = this.data;
+    const newItems = items!.concat(array);
+
+    this.setNewValue(newItems);
+  }
+
+  /**
+   * 	Push item to array in state.
+   *	@public
+   *	@param {Generic} item - Item needs to push
+   */
+  @Action
+  public pushItem(item: T): void {
+    const items = this.data;
+
+    items!.push(item);
+
+    this.setNewValue(items);
+  }
+
+  /**
+   * 	Push item to array in state if not exists in state
+   *	@public
+   *	@param {Generic} item - Item needs to push
+   */
+  @Action
+  public pushItemIfNotExists(item: T): void {
```

Comment:
It looks like some kind of custom action to me.

1) If you want to implement this logic in your specific state, you can create custom action like:

```
@NgxState()
@Injectable()
export class TodoItemsState extends ArrayState<TodoItem[]> {
    public pushIfNotExists(todoItem: TodoItem): void {
        if (
            this.data &&
            this.data.filter((i) => this.compareItems(i, item)).length === 0
        ) {
            this.pushItem(item);
        }
    }
}
```

2) If you want to use this logic often in your project, you can create your own `base state`, like:

```
// Abstract base state
@NgxState()
export abstract class ArrayWithUniqueItemsState<T> extends ArrayState<T> {
    public override pushItem(item: T): void {
        if (
            this.data &&
            this.data.filter((i) => this.compareItems(i, item)).length === 0
        ) {
            super.pushItem(item);
        }
    }
    
    ...
}

// Concrete state
@NgxState()
@Injectable()
export class TodoItemsState extends ArrayWithUniqueItemsState<TodoItem> {}
```